File: tools/python/Cibyl/BinaryTranslation/Mips/mips.py
######################################################################
##
## Copyright (C) 2006,  Simon Kagstrom
##
## Filename:	  Mips.py
## Author:		Simon Kagstrom <ska@bth.se>
## Description:   MIPS defs
##
## $Id: mips.py 12821 2006-12-22 14:46:47Z ska $
##
######################################################################
OP_ADD	= 1
OP_ADDI   = 2
OP_ADDIU  = 3
OP_ADDU   = 4
OP_AND	  = 5
OP_ANDI   = 6
OP_BEQ	= 7
OP_BGEZ   = 8
OP_BGEZAL = 9
OP_BGTZ   = 10
OP_BLEZ   = 11
OP_BLTZ   = 12
OP_BLTZAL = 13
OP_BNE	= 14
OP_DIV	= 16
OP_DIVU   = 17
OP_J	  = 18
OP_JAL	= 19
OP_JALR   = 20
OP_JR	 = 21
OP_LB	 = 22
OP_LBU	= 23
OP_LH	 = 24
OP_LHU	= 25
OP_LUI	= 26
OP_LW	 = 27
OP_LWL	= 28
OP_LWR	= 29
OP_MFHI   = 31
OP_MFLO   = 32
OP_MTHI   = 34
OP_MTLO   = 35
OP_MULT   = 36
OP_MULTU  = 37
OP_NOR	= 38
OP_OR	 = 39
OP_ORI	= 40
OP_RFE	= 41
OP_SB	 = 42
OP_SH	 = 43
OP_SLL	= 44
OP_SLLV   = 45
OP_SLT	= 46
OP_SLTI   = 47
OP_SLTIU  = 48
OP_SLTU   = 49
OP_SRA	= 50
OP_SRAV   = 51
OP_SRL	= 52
OP_SRLV   = 53
OP_SUB	= 54
OP_SUBU   = 55
OP_SW	 = 56
OP_SWL	= 57
OP_SWR	= 58
OP_XOR	= 59
OP_XORI   = 60
OP_SYSCALL= 61
OP_UNIMP  = 62
OP_RES	= 63
OP_MFCP0  = 65
OP_MTCP0  = 66
OP_CFCP0  = 67
OP_CTCP0  = 68
OP_BCCP0  = 69
OP_TLBR   = 70
OP_TLBWI  = 71
OP_TLBWR  = 72
OP_TLBP   = 73
OP_COP1   = 75
OP_COP2   = 76
OP_COP3   = 77
OP_BREAK  = 78

registerNames = {
	 0 : "zero",
	 1 : "at",
	 2 : "v0",
	 3 : "v1",
	 4 : "a0",
	 5 : "a1",
	 6 : "a2",
	 7 : "a3",
	 8 : "t0",
	 9 : "t1",
	10 : "t2",
	11 : "t3",
	12 : "t4",
	13 : "t5",
	14 : "t6",
	15 : "t7",
	16 : "s0",
	17 : "s1",
	18 : "s2",
	19 : "s3",
	20 : "s4",
	21 : "s5",
	22 : "s6",
	23 : "s7",
	24 : "t8",
	25 : "t9",
	26 : "k0",
	27 : "k1",
	28 : "gp",
	29 : "sp",
	30 : "s8", # or fp
	31 : "ra",
	32 : "hi", # special
	33 : "lo",

	# Not part of the regular instruction set
	34 : "cpc", # Cibyl cross-method PC
	35 : "cm0", # Cibyl memory access registers
	36 : "cm1",
	37 : "cm2",
	38 : "cm3",
	39 : "cm4",
	40 : "cm5",
	41 : "cm6",
	42 : "cm7",
	43 : "cm8",
	44 : "ecb", # Exception callback address
	45 : "ear", # Exception argument
	46 : "fna", # Function address (for multiple java methods)
	47 : "mem", # A reference to the memory
}

# An inversion of the above
namesToRegisters = dict([ (v, k) for k, v in registerNames.iteritems() ])

memoryAddressRegisters = [
	namesToRegisters["cm0"],
	namesToRegisters["cm1"],
	namesToRegisters["cm2"],
	namesToRegisters["cm3"],
	namesToRegisters["cm4"],
	namesToRegisters["cm5"],
	namesToRegisters["cm6"],
	namesToRegisters["cm7"],
	namesToRegisters["cm8"],
]

callerSavedRegisters = [
	namesToRegisters["s0"],
	namesToRegisters["s1"],
	namesToRegisters["s2"],
	namesToRegisters["s3"],
	namesToRegisters["s4"],
	namesToRegisters["s5"],
	namesToRegisters["s6"],
	namesToRegisters["s7"],
	namesToRegisters["s8"],
]


R_ZERO = namesToRegisters["zero"]
R_V0 = namesToRegisters["v0"]
R_V1 = namesToRegisters["v1"]
R_A0 = namesToRegisters["a0"]
R_A1 = namesToRegisters["a1"]
R_A2 = namesToRegisters["a2"]
R_A3 = namesToRegisters["a3"]
R_SP = namesToRegisters["sp"]
R_RA = namesToRegisters["ra"]
R_HI = namesToRegisters["hi"]
R_LO = namesToRegisters["lo"]
R_K1 = namesToRegisters["k1"]
R_CPC= namesToRegisters["cpc"]
R_ECB= namesToRegisters["ecb"] # Exception callback address
R_EAR= namesToRegisters["ear"] # Exception argument
R_FNA= namesToRegisters["fna"] # Function address (for multiple functions / Java method)
R_MEM= namesToRegisters["mem"]

N_REGS = len(namesToRegisters)

ifmtZeroExtend = [
	OP_ANDI,
	OP_ORI,
	OP_XORI,
]

delaySlotInstructions = [
	# Conditional jumps
	OP_BEQ,
	OP_BGEZ,
	OP_BGEZAL,
	OP_BGTZ,
	OP_BLEZ,
	OP_BLTZ,
	OP_BLTZAL,
	OP_BNE,
	# Unconditional jumps
	OP_J,
	OP_JAL,
	OP_JALR,
	OP_JR,
	# Load delay slots
#	OP_LB,
#	OP_LBU,
#	OP_LH,
#	OP_LHU,
#	OP_LW,
#	OP_LWL,
#	OP_LWR,
]

#The table below is used to translate bits 31:26 of the instruction
#into a value suitable for the "opCode" field of a MemWord structure,
#or into a special value for further decoding.
SPECIAL  = 100
BCOND = 101
COP0 = 102
COP1 = 103
COP2 = 104
COP3 = 105
CIBYL_SYSCALL  = 999
CIBYL_REGISTER_ARGUMENT  = 1000
CIBYL_ASSIGN_MEMREG = 1001

IFMT = 1
JFMT = 2
RFMT = 3

OPCODE = 0
FORMAT = 1

opTable = [
	[SPECIAL, RFMT], [BCOND, IFMT], [OP_J, JFMT], [OP_JAL, JFMT],
	[OP_BEQ, IFMT], [OP_BNE, IFMT], [OP_BLEZ, IFMT], [OP_BGTZ, IFMT],
	[OP_ADDI, IFMT], [OP_ADDIU, IFMT], [OP_SLTI, IFMT], [OP_SLTIU, IFMT],
	[OP_ANDI, IFMT], [OP_ORI, IFMT], [OP_XORI, IFMT], [OP_LUI, IFMT],
	[COP0, RFMT], [COP1, IFMT], [COP2, IFMT], [COP3,IFMT],
	[OP_RES, IFMT], [OP_RES, IFMT], [OP_RES, IFMT], [OP_RES, IFMT],
	[OP_RES, IFMT], [OP_RES, IFMT], [OP_RES, IFMT], [OP_RES, IFMT],
	[OP_RES, IFMT], [OP_RES, IFMT], [OP_RES, IFMT], [OP_RES, IFMT],
	[OP_LB, IFMT], [OP_LH, IFMT], [OP_LWL, IFMT], [OP_LW, IFMT],
	[OP_LBU, IFMT], [OP_LHU, IFMT], [OP_LWR, IFMT], [OP_RES, IFMT],
	[OP_SB, IFMT], [OP_SH, IFMT], [OP_SWL, IFMT], [OP_SW, IFMT],
	[OP_RES, IFMT], [OP_RES, IFMT], [OP_SWR, IFMT], [OP_RES, IFMT],
	[OP_UNIMP, IFMT], [OP_UNIMP, IFMT], [OP_UNIMP, IFMT], [OP_UNIMP,IFMT],
	[OP_RES, IFMT], [OP_RES, IFMT], [OP_RES, IFMT], [OP_RES, IFMT],
	[OP_UNIMP, IFMT], [OP_UNIMP, IFMT], [OP_UNIMP, IFMT], [OP_UNIMP,IFMT],
	[OP_RES, IFMT], [OP_RES, IFMT], [OP_RES, IFMT], [OP_RES, IFMT]
	]

specialTable = [
	OP_SLL, OP_RES, OP_SRL, OP_SRA, OP_SLLV, OP_RES, OP_SRLV, OP_SRAV,
	OP_JR, OP_JALR, OP_RES, OP_RES, OP_SYSCALL, OP_BREAK, OP_RES, OP_RES,
	OP_MFHI, OP_MTHI, OP_MFLO, OP_MTLO, OP_RES, OP_RES, OP_RES, OP_RES,
	OP_MULT, OP_MULTU, OP_DIV, OP_DIVU, OP_RES, OP_RES, OP_RES, OP_RES,
	OP_ADD, OP_ADDU, OP_SUB, OP_SUBU, OP_AND, OP_OR, OP_XOR, OP_NOR,
	OP_RES, OP_RES, OP_SLT, OP_SLTU, OP_RES, OP_RES, OP_RES, OP_RES,
	OP_RES, OP_RES, OP_RES, OP_RES, OP_RES, OP_RES, OP_RES, OP_RES,
	OP_RES, OP_RES, OP_RES, OP_RES, OP_RES, OP_RES, OP_RES, OP_RES
	]

opStrings = [
	"Shouldn't happen", "add  ", "addi ", "addiu", "addu ", "and  ",
	"andi ", "beq  ", "bgez ", "bgezal", "bgtz ", "blez ", "bltz ",
	"bltzal", "bne  ", "Shouldn't happen", "div  ", "divu ", "j	",
	"jal  ", "jalr ", "jr   ", "lb   ", "lbu  ", "lh   ", "lhu  ",
	"lui  ", "lw   ", "lwl  ", "lwr  ", "Shouldn't happen", "mfhi ",
	"mflo ", "shouldn't happen", "mthi ", "mtlo ", "mult ","multu",
	"nor  ", "or   ", "ori  ", "rfe  ", "sb   ", "sh   ", "sll  ",
	"sllv ","slt  ", "slti ", "sltiu", "sltu ", "sra  ", "srav ",
	"srl  ", "srlv ", "sub  ", "subu ", "sw   ", "swl  ", "swr  ",
	"xor  ", "xori ", "syscall", "Unimplemented", "Reserved",
	"Shouldn't happen", "mfcp0", "mtcp0", "cfcp0", "ctcp0", "bccp0",
	"tlbr ", "tlbwi", "tlbwr", "tlbp ", "Shouldn't happen", "??CP1",
	"??CP2", "??CP3", "break"
	]

File: tools/compiled/translator/mips.cc
/*********************************************************************
 *
 * Copyright (C) 2008,  Simon Kagstrom
 *
 * Filename:      mips.cc
 * Author:        Simon Kagstrom <simon.kagstrom@gmail.com>
 * Description:   MIPS implementation
 *
 * $Id:$
 *
 ********************************************************************/
#include <mips.hh>

mips_op_entry_t mips_op_entries[] =
{
  {SPECIAL, RFMT}, {BCOND, IFMT},    {OP_J, JFMT},    {OP_JAL, JFMT},
  {OP_BEQ, IFMT},  {OP_BNE, IFMT},   {OP_BLEZ, IFMT}, {OP_BGTZ, IFMT},
  {OP_ADDI, IFMT}, {OP_ADDIU, IFMT}, {OP_SLTI, IFMT}, {OP_SLTIU, IFMT},
  {OP_ANDI, IFMT}, {OP_ORI, IFMT},   {OP_XORI, IFMT}, {OP_LUI, IFMT},
  {COP0, RFMT},    {COP1, IFMT},     {COP2, IFMT},    {COP3,IFMT},
  {OP_RES, IFMT},  {OP_RES, IFMT},   {OP_RES, IFMT},  {OP_RES, IFMT},
  {OP_RES, IFMT},  {OP_RES, IFMT},   {OP_RES, IFMT},  {OP_RES, IFMT},
  {OP_RES, IFMT},  {OP_RES, IFMT},   {OP_RES, IFMT},  {OP_RES, IFMT},
  {OP_LB, IFMT},   {OP_LH, IFMT},    {OP_LWL, IFMT},  {OP_LW, IFMT},
  {OP_LBU, IFMT},  {OP_LHU, IFMT},   {OP_LWR, IFMT},  {OP_RES, IFMT},
  {OP_SB, IFMT},   {OP_SH, IFMT},    {OP_SWL, IFMT},  {OP_SW, IFMT},
  {OP_RES, IFMT},  {OP_RES, IFMT},   {OP_SWR, IFMT},  {OP_RES, IFMT},
  {OP_UNIMP, IFMT},{OP_LWC1, IFMT}, {OP_UNIMP, IFMT},{OP_UNIMP,IFMT},
  {OP_RES, IFMT},  {OP_RES, IFMT},   {OP_RES, IFMT},  {OP_RES, IFMT},
  {OP_UNIMP, IFMT},{OP_SWC1, IFMT}, {OP_UNIMP, IFMT},{OP_UNIMP,IFMT},
  {OP_RES, IFMT},  {OP_RES, IFMT},   {OP_RES, IFMT},  {OP_RES, IFMT}
};

mips_opcode_t mips_special_table[] =
{
  OP_SLL,  OP_RES,   OP_SRL,  OP_SRA,  OP_SLLV,    OP_RES,   OP_SRLV, OP_SRAV,
  OP_JR,   OP_JALR,  OP_RES,  OP_RES,  OP_SYSCALL, OP_BREAK, OP_RES,  OP_RES,
  OP_MFHI, OP_MTHI,  OP_MFLO, OP_MTLO, OP_RES,     OP_RES,   OP_RES,  OP_RES,
  OP_MULT, OP_MULTU, OP_DIV,  OP_DIVU, OP_RES,     OP_RES,   OP_RES,  OP_RES,
  OP_ADD,  OP_ADDU,  OP_SUB,  OP_SUBU, OP_AND,     OP_OR,    OP_XOR,  OP_NOR,
  OP_RES,  OP_RES,   OP_SLT,  OP_SLTU, OP_RES,     OP_RES,   OP_RES,  OP_RES,
  OP_RES,  OP_RES,   OP_RES,  OP_RES,  OP_RES,     OP_RES,   OP_RES,  OP_RES,
  OP_RES,  OP_RES,   OP_RES,  OP_RES,  OP_RES,     OP_RES,   OP_RES,  OP_RES
};

mips_opcode_t mips_cop1_table[] =
{
  OP_FADD, OP_FSUB, OP_FMUL, OP_FDIV, OP_FSQRT, OP_FABS, OP_FMOV, OP_FNEG,
  OP_ROUND_L, OP_TRUNC_L, OP_CEIL_L, OP_FLOOR_L, OP_ROUND_W, OP_TRUNC_W, OP_CEIL_W,
  OP_FLOOR_W,
  /* 16-31, reserved */
  OP_RES, OP_RES, OP_RES, OP_RES, OP_RES, OP_RES, OP_RES, OP_RES, OP_RES,
  OP_RES, OP_RES, OP_RES, OP_RES, OP_RES, OP_RES, OP_RES,
  OP_CVT_S, OP_CVT_D, OP_RES, OP_RES, OP_CVT_W, OP_CVT_L,
  /* 38-47, reserved */
  OP_RES, OP_RES, OP_RES, OP_RES, OP_RES, OP_RES, OP_RES, OP_RES, OP_RES,
  /* 48-63, float compares */
  OP_C_F, OP_C_UN, OP_C_EQ, OP_C_UEQ, OP_C_OLT, OP_C_ULT, OP_C_OLE, OP_C_ULE,
  OP_C_SF, OP_C_NGLE, OP_C_SEQ, OP_C_NGL, OP_C_LT, OP_C_NGE, OP_C_LE, OP_C_NGT
};

const char *mips_op_strings[] =
{
  "Shouldn't happen", "add  ", "addi ", "addiu", "addu ", "and  ",
  "andi ", "beq  ", "bgez ", "bgezal", "bgtz ", "blez ", "bltz ",
  "bltzal", "bne  ", "Shouldn't happen", "div  ", "divu ", "j	",
  "jal  ", "jalr ", "jr   ", "lb   ", "lbu  ", "lh   ", "lhu  ",
  "lui  ", "lw   ", "lwl  ", "lwr  ", "Shouldn't happen", "mfhi ",
  "mflo ", "shouldn't happen", "mthi ", "mtlo ", "mult ","multu",
  "nor  ", "or   ", "ori  ", "rfe  ", "sb   ", "sh   ", "sll  ",
  "sllv ","slt  ", "slti ", "sltiu", "sltu ", "sra  ", "srav ",
  "srl  ", "srlv ", "sub  ", "subu ", "sw   ", "swl  ", "swr  ",
  "xor  ", "xori ", "syscall", "Unimplemented", "Reserved",
  "Shouldn't happen", "mfcp0", "mtcp0", "cfcp0", "ctcp0", "bccp0",
  "tlbr ", "tlbwi", "tlbwr", "tlbp ", "Shouldn't happen", "??CP1",
  "??CP2", "??CP3", "break",
  /* FPU */
  "mfc1", "cfc1", "mtc1", "ctc1", "bc1f", "bc1t", "lwc1", "swc1",
  "fadd", "fsub", "fmul", "fdiv", "fsqrt", "fabs", "fmov", "fneg",
  "round.l", "trunc.l", "ceil.l", "floor.l", "ceil.w", "floor.w",
  "round.w", "trunc.w", "cvt.s", "cvt.d", "cvt.w", "cvt.l",
  "Reserved","Reserved","Reserved","Reserved","Reserved",
  "Reserved","Reserved","Reserved","Reserved",
  "c.f", "c.un", "c.eq", "c.ueq", "c.olt", "c.ult", "c.ole", "c.ule",
  "c.sf", "c.ngle", "c.seq", "c.ngl", "c.lt", "c.nge", "c.le", "c.ngt"
};

const char *mips_reg_strings[] =
{
  "zero", "at", "v0", "v1", "a0", "a1", "a2", "a3", "t0", "t1", "t2", "t3", "t4", "t5",
  "t6", "t7", "s0", "s1", "s2", "s3", "s4", "s5", "s6", "s7", "t8", "t9", "k0", "k1",
  "gp", "sp", "s8", "ra", "hi", "lo",
  "RES","RES","RES","RES","RES","RES",
  "f0", "f1", "f2", "f3", "f4", "f5", "f6", "f7", "f8", "f9", "f10", "f11", "f12", "f13",
  "f14", "f15", "f16", "f17", "f18", "f19", "f20", "f21", "f22", "f23", "f24", "f25",
  "f26", "f27", "f28", "f29", "f30", "f31",
  "cpc", "cm0", "cm1", "cm2", "cm3", "cm4", "cm5", "cm6", "cm7", "madr", "ecb", "ear",
  "fna", "mem",
};

MIPS_register_t mips_caller_saved[] = { R_S0, R_S1, R_S2, R_S3, R_S4, R_S5, R_S6, R_S7, R_S8, R_ZERO };

static bool ifmt_zero_extend[ N_INSNS ];

static void __attribute__((constructor))init()
{
  ifmt_zero_extend[ OP_ANDI ] = true;
  ifmt_zero_extend[ OP_ORI ]  = true;
  ifmt_zero_extend[ OP_XORI ] = true;
}

bool mips_zero_extend_opcode(mips_opcode_t opcode)
{
  if (opcode > N_INSNS)
    return false;
  return ifmt_zero_extend[opcode];
}
